### DISCUSSION TITLE
Call logs & tlms in an object-oriented programing way

### DISCUSSION THREAD

**Discussion by kitty99099** [nasa/fprime#2889] [Q&A] [answered]
Hello everyone.
I have been worked with F' for about 3 months to implement F' into a commercial satellite subsystem. My project is near 100% successful, but there came out to be a little problem that I cannnot go to the next step. It is a programing skill problem, so I'd like to see if there is anyone also met the same problem or anyone know how to fix the problem.
The thing is:
There are `component.fpp`, `component.hpp` and `component.cpp` files in the component based on F' framework, right? I implement many functions in the .cpp file. To prevent my component.cpp from becoming untidy and unreadable. I used the OOP way to make some classes including some functions and varible on another file that my component.cpp can easily use these packaged functions and variables. Everything is going well until in the last step......There are so many packaged functions in the class built by F' like: `this->tlmWrite_XXX(OOO)`, `this->log_XXX(OOO)`. These two functions let us to show the telemetry channel and event on the GDS. However there are so many telemetry and event I want to show. If I write all of these in the component.cpp, it will become very untidy and unreadable which I don't want. Thus, I tried to use another file to inherit from the F' auto-generated class, but it is really hard to use because F' also auto-generate many other virtual function. If I want to inherit from the F' auto-generated class, I need to implement all of these virtual functions first, and then I can use `this->tlmWrite_XXX(OOO)`, `this->log_XXX(OOO)`......
My programing experience is not rich as many people here, so I have been stuck here for a while. Is there any recommended solution? I appreciate to any of your kind help.

(Below is an example to show what I don't want.)
in `component.cpp`
```
if (a) {
    this->tlmWrite_A(a)
}
if (b) {
    this->tlmWrite_B(b)
}
.
.
.
.
.
.
if (x) {
    this->log_X(x)
}
if (y) {
    this->log_Y(y)
}
.
.
.
.
.
.
```

**Comment by thomas-bc** [comment, marked as answer]
If it were me I would write these helper functions you want within `component.cpp`.
Longer file doesn't mean untidy code; I do understand where you're coming from in terms of trying to compartmentalize things but these telemetry functions are very much belonging to the component.

I would definitely recommend **against** inheriting from the F´ autocoded base classes.

**Reply by thomas-bc** [reply]
Here's an example in the Svc.Version component ([here](https://github.com/nasa/fprime/blob/dfc655f7892802e5ecfcd9e9a9c133c76ef2aa59/Svc/Version/Version.cpp#L187)) where we have a helper function to emit telemtry.

**Reply by kitty99099** [reply]
Dear @thomas-bc,
Thank you for your kind and fast answer. What I wanted to do is like the Svc.Version you gave, "switch cases". However, I have more than 50 telemetries need to be shown. That's why I was afraid of the component.cpp became very long, and then I tried to write the "switch cases" in another file by an OOP way. After your explanation, I feel better and relax that a long file doesn't mean anything. Also, I realized that maybe write these code directly in the component.cpp is the best way to perform F' framework.

**Comment by LeStarch** [comment]
Another possible solution is to use a dispatch table.  Essentially, you can use an array of function pointers to the telemetry functions.  Here is some pseudo code

```
m_array = [
   &this->tlmWrite_XXX,
   &this->tlmWrite_XXY,
... 
]
```

You then cal call these functions in a for loop:

```
for (int i = 0; i < size of array; i++) {
    m_array[i](000);
}
```

We are aware of the problems with large numbers of telemetry, and have it on our list to work on.  For now, this solution, or @thomas-bc's are work arounds!

**Reply by kitty99099** [reply]
Dear @LeStarch,
Thank you for your kind help, too. You show me a great way to make my code tidier. I appreciate your kind answer even though what I asked seems like just a programing skill problem.

**Comment by kitty99099** [comment]
Dear @thomas-bc and @LeStarch,
Thank you once again for your response. I am delighted to receive an answer directly from NASA JPL in such a short time.
BTW, thank you all to develop and maintain F'. I use F' to implement an communication within an OBC and an ADCS, so that I just won an opportunity to go to the USA and to visit NASA JPL in this November! I am so exited that my dream will come true!
